index.js

```diff
@@ -37,41 +38,163 @@ var EchoClient = function(config){
  * @param {object} data.props Other properties associated with the event
  * @callback callback
  */
-EchoClient.prototype.addEvents = function(token, data, callback){
-    if(!token){
+ EchoClient.prototype.addEvents = function(token, data, callback){
+     if(!token){
+         throw new Error('Missing Persona token');
+     }
+
+     if(!data){
+         throw new Error('Missing data');
+     }
+
+     // multiple events can be written by posting an array
+     // if not an array, check data looks ok
+     if (!(data instanceof Array)) {
+         if(!data.class){
+             throw new Error('Missing field data.class');
+         }
+         if(!data.source){
+             throw new Error('Missing field data.source');
+         }
+     }
+
+     var requestOptions = {
+         url: this.config.echo_endpoint + '/1/events',
+         headers: {
+             'Accept': 'application/json',
+             'Authorization': 'Bearer ' + token
+         },
+         body: data,
+         method: 'POST',
+         json: true
+     };
+
+     this.debug(JSON.stringify(requestOptions));
+
+     request.post(requestOptions, function(err, response, body){
+         if(err){
+             callback(err);
+         } else{
+             callback(null, body);
+         }
+     });
+ };
+
+/**
+ * Query analytics using a passed operator and parameters
+ * @param  {string}   token             Persona token
+ * @param  {string}   queryOperator     Query operator (hits, average, sum, max or funnel)
+ * @param  {object}   queryParams       Hash of parameters to add to the query
+ * @param  {boolean}  useCache          Indicates if cache should be used or not
+ * @callback callback
+ */
+EchoClient.prototype.queryAnalytics = function(token, queryOperator, queryParams, useCache, callback) {
+    if (!token) {
         throw new Error('Missing Persona token');
     }
-    if(!data){
-        throw new Error('Missing data');
+
+    if (!queryOperator) {
+        throw new Error('Missing Analytics queryOperator');
+    }
+
+    var validOperators = ['hits', 'average', 'sum', 'max', 'funnel'];
+
+    if (validOperators.indexOf(queryOperator) === -1) {
+        throw new Error('Invalid Analytics queryOperator');
     }
-    if(!data.class){
-        throw new Error('Missing field data.class');
+
+    if (!queryParams) {
+        throw new Error('Missing Analytics queryParams');
     }
-    if(!data.source){
-        throw new Error('Missing field data.source');
+
+    var constructQueryStringResponse = this._queryStringParams(queryParams);
+
+    if (constructQueryStringResponse.errors) {
```

Comment:
There's quite a bit of work to get these errors in the function `isValidParameter` further down yet you're not utilising that information here.